### PR TITLE
Feature/backend data transmission

### LINKED ISSUE #4: FastAPI setup for fastcrawler
- [x] install fastcrawler core
- [x] Connect fastcrawler's core to fastapi
- [x] Implement task controller as backend route in fastapi
- [x] write tests to check behaviour of controller

### LINKED ISSUE #7: Task Router
- [x] Router to return task, set them, and control them.

### PR DESCRIPTION
**Pull request by ManiMozaffar**
closes #4 
closes #7

### CODE REVIEW

File: fastcrawler_ui/routers/spiders.py [hunk 1]
```diff
@@ -1,12 +1,30 @@
 from fastapi import APIRouter, Depends
 from fastcrawler import FastCrawler
 
+from fastcrawler_ui.controllers.spider import CrawlerController
 from fastcrawler_ui.core.fastapi.depends import get_crawler
+from fastcrawler_ui.repository.spiders import CrawlerRepository
 
 router = APIRouter()
+crawler_repository = CrawlerRepository()
```
Comment (ManiMozaffar) [inline fastcrawler_ui/routers/spiders.py]: Don't use global variables in router. initiate it inside the controller if you have to.

File: fastcrawler_ui/routers/spiders.py [hunk 2]
```diff
@@ -1,12 +1,30 @@
...
+
+
+# @router.get("/all")
+# async def clients(crawler: FastCrawler = Depends(get_crawler)):
+#     results = await crawler.controller.all()
+#     return [result.model_dump() for result in results]
 
 
 @router.get("/all")
 async def clients(crawler: FastCrawler = Depends(get_crawler)):
```
Comment (ManiMozaffar) [inline fastcrawler_ui/routers/spiders.py]: write test for each router and make sure it doesn't fail.

File: fastcrawler_ui/repository/spiders.py [hunk 3]
```diff
@@ -117,23 +117,22 @@ async def update_task(
...
+    def get_tasks_in_processes(self, crawler: FastCrawler) -> dict[Process, Task]:
         """Get all tasks from the crawler.
 
         Returns:
             Dict[Process, Task]
         """
         self.crawlers.add(crawler)
-        processes = {process: process.task for process in crawler.crawlers}
+        processes = {process: task for process, task in zip(crawler.crawlers, crawler.controller.app.task_lib.session.tasks)}
         return processes
```
Comment (aerosadegh) [inline fastcrawler_ui/repository/spiders.py:147]: get tasks from internal (rocketry) instead of crawlers

Comment (aerosadegh) [reply fastcrawler_ui/repository/spiders.py:147]: there is a bug that fixed in  [3c10614](https://github.com/fast-crawler/ui/pull/6/commits/3c10614d59fb1538e1a1654e9c0adff2580220d5)

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    BaseModel,
+    Depends,
+    FastCrawler as _FastCrawler,
+    Process,
+    Spider,
+    XPATHField,
+)
+
+
+sys.path.append(os.path.dirname(os.path.dirname(__file__)))
```
Comment (ManiMozaffar) [inline tests/conftest.py]: this should not be inside conftest. 
When you're using pytest, you have the access level to define modules within the pytest env, which is nicer and code editor would still detect type anno and you can jump to the impl from code editor.

File: tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    async def save_cycle(self, all_data: list[RequestCycle]) -> None:
+        global stopped_crawler_flag
+        stopped_crawler_flag += 1
+        return await super().save_cycle(all_data)
+
+    async def save(self, all_data: list[PersonPage]):
+        global stopped_crawler_flag
+        assert all_data is not None
+        assert len(all_data) == 10
+        assert stopped_crawler_flag == 1
```
Comment (ManiMozaffar) [inline tests/conftest.py]: Do not assert in conftest. if it's a test assertion, then there has to be a test for this.

File: tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                spider=MySpider(),
+                cond="every 5 minute",
+            ),
+        )
+    )
+    return crawler
+
+
+@pytest_asyncio.fixture(scope="session")
+def client():
```
Comment (ManiMozaffar) [inline tests/conftest.py]: when you're writing fixture in pytest, make sure you're including their return type, so we know that what has been used in fixture, to copy/paste the fixture within our codebase.

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+
+@pytest.fixture
+def manager():
+    yield ConnectionRepository()
+
+
+@pytest.fixture
+def websocket():
+    yield AsyncMock()
```
Comment (ManiMozaffar) [inline tests/conftest.py:91]: this fixture is used nowhere.

File: tests/test_managers.py [hunk 8]
```diff
@@ -1,62 +1,61 @@
...
+def test_websocket_connect(manager: ConnectionRepository, websocket):
+    asyncio.run(manager.connect(websocket))
+    assert websocket in manager.active_connections
 
-        websocket_mock1.send_json.assert_called_once_with(message.model_dump())
-        websocket_mock2.send_json.assert_called_once_with(message.model_dump())
 
-    def test_connection_manager_send_personal_message(self):
-        websocket_mock = AsyncMock()
+def test_connection_manager_broadcast(manager: ConnectionRepository):
```
Comment (ManiMozaffar) [inline tests/test_managers.py]: the tests are still written with unittest. Please get rid of AsyncMock.
Instead mock the WSController.

File: tests/test_managers.py [hunk 9]
```diff
@@ -1,62 +1,61 @@
...
+    manager.active_connections.add(websocket_mock1)
+    manager.active_connections.add(websocket_mock2)
 
-        asyncio.run(self.manager.send_personal_message(
-            message, websocket_mock))
+    message_dict = {"content": "Hello", "sender": "Sadegh"}
+    message = Message(**message_dict)
 
-        websocket_mock.send_json.assert_called_once_with(message.model_dump())
+    asyncio.run(manager.broadcast(message))
```
Comment (ManiMozaffar) [inline tests/test_managers.py]: use async function, don't use asyncio.run(...) in pytest. Pytest has its own event loop strategy.

File: tests/test_managers.py [hunk 10]
```diff
@@ -1,62 +1,61 @@
...
+    asyncio.run(manager.send_personal_message(message, websocket_mock))
+
+    websocket_mock.send_json.assert_called_once_with(message.model_dump())
+
+
+def test_disconnects(manager: ConnectionRepository):
+    websocket = AsyncMock()
+    websocket2 = AsyncMock()
+    websocket3 = AsyncMock()
+    asyncio.run(manager.connect(websocket))
```
Comment (ManiMozaffar) [inline tests/test_managers.py]: Same applye here.

File: tests/test_ws.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+# import unittest
+# from unittest.mock import patch
+
+# from fastapi.testclient import TestClient
+# from main import app
+
+
+# class WebSocketEndpointTests(unittest.TestCase):
+#     @patch("main.websocket_endpoint")
+#     def test_websocket_endpoint(self, websocket_endpoint_mock):
```
Comment (ManiMozaffar) [inline tests/test_ws.py:12]: Remove commented file from source code if you intend to deprecate it.

File: fastcrawler_ui/controllers/ws.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class WSController:
+    def __init__(self, ws_connection_repo: ConnectionRepository = Depends(ConnectionRepository)):
+        self.ws_connection_repo = ws_connection_repo
+
+    async def get_connection(self, websocket: WebSocket):
+        await self.ws_connection_repo.connect(websocket)
+        try:
+            while True:
```
Comment (ManiMozaffar) [inline fastcrawler_ui/controllers/ws.py:15]: Decouple the condition to another method. Monkey patch that method when you're writing tests in your mock class

File: fastcrawler_ui/controllers/spider.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            task_name=task_name,
+        )
+        await process.start()
+
+    async def stop_task_by_name(self, crawler: FastCrawler, task_name: str):
+        task = await self.get_process_by_task(
+            crawler=crawler,
+            task_name=task_name,
+        )
+        await task.stop()
```
Comment (ManiMozaffar) [inline fastcrawler_ui/controllers/spider.py]: You should not use Spider to disable the task
Use the Process to disable a crawling process
Go in process.py, where class Process is defined. there is stop and start method, which works great.

That logic is written to ensure data integrity, and stop the process even if it's in middle of crawling. The request batch has to be performed, then on next batch, the while condition fails and hence the crawler start.

Remember that if you stop a crawler, it is stopped forever!
and if you start a crawler, it's stopped in rocketry. Maybe you should use background process for start. but stop is responding fast.

**Pull request merged**